SourceCode/h8/light.cpp

```diff
@@ -0,0 +1,796 @@
+        return fields;
+    }
+
+    void setRows(const unsigned int rows) {
+        this->rows = rows;
+    }
+    void setColumns(const unsigned int columns) {
+        this->columns = columns;
+    }
+    void setField(const unsigned int i, const unsigned int j, const float value) {
+        fields[i][j] = value;
+    }
+    void setFields(const float * const matrix) {
+        for(unsigned int i = 0; i < rows; i++) {
+            for(unsigned int j = 0; j < columns; j++) {
+                fields[i][j] = matrix[j + columns * i];
+            }
+        }
+    }
+
+    void print() const {
+        for(unsigned int i = 0; i < rows; i++) {
+            for(unsigned int j = 0; j < columns; j++) {
+                std::cout << fields[i][j] << " ";
+            }
+            std::cout << std::endl;
+        }
+    }
+};
+
+class Vector : public Matrix {
+public:
+    Vector(const float x, const float y, const float z) 
+        : Matrix(std::begin({x, y, z}), 1, 3) 
+        {}
+    Vector(const Matrix &m) 
+        : Matrix(
+            std::begin({m.getField(0, 0), m.getField(0, 1), m.getField(0, 2)}), 
+            1, 
+            3
+        )
+        {}
+    Vector(const float n) 
+        : Matrix(std::begin({n, n, n}), 1, 3) 
+        {}
+
+    float length() const {
+        return sqrt(getField(0, 0) * getField(0, 0) 
+                        + getField(0, 1) * getField(0, 1)
+                        + getField(0, 2) * getField(0, 2));
+    }
+    void normalize() {
+        float cachedLength = length();
+        cachedLength = 1 / cachedLength;
+        setField(0, 0, getField(0, 0) * cachedLength);
+        setField(0, 1, getField(0, 1) * cachedLength);
+        setField(0, 2, getField(0, 2) * cachedLength);
+    }
+
+    float getX() const {
+        return getField(0, 0);
+    }
+    float getY() const {
+        return getField(0, 1);
+    }
+    float getZ() const {
+        return getField(0, 2);
+    }
+    float get(const unsigned int index) const {
+        return getField(0, index);
+    }
+
+    void setX(const float x) {
+        setField(0, 0, x);
+    }
+    void setY(const float y) {
+        setField(0, 1, y);
+    }
+    void setZ(const float z) {
+        setField(0, 2, z);
+    }
+    void set(const unsigned int index, const float value) {
+        setField(0, index, value);
+    }
+
+    void print() const {
+        Matrix::print();
+    }
+
+    Vector &operator=(const Matrix& m) {
+        for(unsigned int i = 0; i < 3; i++) {
+            setField(0, i, m.getField(0, i));
+        }
+        return *this;
+    }
+    Vector &operator=(const float n) {
+        for(unsigned int i = 0; i < 3; i++) {
+            set(i, n);
+        }
+        return *this;
+    }
+
+    Vector operator*(const Vector& v) {
+        Vector result(0, 0, 0);
+        for(unsigned int i = 0; i < 3; i++) {
+            result.set(i, get(i) * v.get(i));
+        }
+        return result;
+    }
+};
+
+namespace Math {
```

Comment:
I not see matrix vector multiplication operator?

Reply:
Mathematically speaking a Vector is a special case of a matrix (1x3 or 3x1). In addition, it allows me to use matrix operations and multiplications on vector without code duplication. @VladislavVulchevChaos